hist/histv7/inc/ROOT/RHistStats.hxx

```diff
@@ -0,0 +1,351 @@
+
+#ifndef ROOT_RHistStats
+#define ROOT_RHistStats
+
+#include "RHistUtils.hxx"
+#include "RLinearizedIndex.hxx"
+#include "RWeight.hxx"
+
+#include <cmath>
+#include <cstdint>
+#include <stdexcept>
+#include <tuple>
+#include <vector>
+
+class TBuffer;
+
+namespace ROOT {
+namespace Experimental {
+
+/**
+Histogram statistics of unbinned values.
+
+Every call to \ref Fill(const A &... args) "Fill" updates sums to compute the number of effective entries as well as the
+arithmetic mean and other statistical quantities per dimension:
+\code
+ROOT::Experimental::RHistStats stats(1);
+stats.Fill(8.5);
+stats.Fill(1.5);
+// stats.ComputeMean() will return 5.0
+\endcode
+
+\warning This is part of the %ROOT 7 prototype! It will change without notice. It might trigger earthquakes.
+Feedback is welcome!
+*/
+class RHistStats final {
+public:
+   /// Statistics for one dimension.
+   struct DimensionStats final {
+      double fSumWX = 0.0;
+      double fSumWX2 = 0.0;
+      double fSumWX3 = 0.0;
+      double fSumWX4 = 0.0;
+
+      void Add(double x)
+      {
+         fSumWX += x;
+         fSumWX2 += x * x;
+         fSumWX3 += x * x * x;
+         fSumWX4 += x * x * x * x;
+      }
+
+      void Add(double x, double w)
+      {
+         fSumWX += w * x;
+         fSumWX2 += w * x * x;
+         fSumWX3 += w * x * x * x;
+         fSumWX4 += w * x * x * x * x;
+      }
+   };
+
+private:
+   /// The number of entries
+   std::uint64_t fNEntries = 0;
+   /// The sum of weights
+   double fSumW = 0.0;
+   /// The sum of weights squared
+   double fSumW2 = 0.0;
+   /// The sums per dimension
+   std::vector<DimensionStats> fDimensionStats;
+
+public:
+   /// Construct a statistics object.
+   ///
+   /// \param[in] nDimensions the number of dimensions, must be > 0
+   explicit RHistStats(std::size_t nDimensions)
+   {
+      if (nDimensions == 0) {
+         throw std::invalid_argument("nDimensions must be > 0");
+      }
+      fDimensionStats.resize(nDimensions);
+   }
+
+   std::size_t GetNDimensions() const { return fDimensionStats.size(); }
+
+   std::uint64_t GetNEntries() const { return fNEntries; }
+   double GetSumW() const { return fSumW; }
+   double GetSumW2() const { return fSumW2; }
+
+   const DimensionStats &GetDimensionStats(std::size_t dim = 0) const { return fDimensionStats.at(dim); }
+
+   /// Compute the number of effective entries.
+   ///
+   /// \f[
+   /// \frac{(\sum w_i)^2}{\sum w_i^2}
+   /// \f]
+   ///
+   /// \return the number of effective entries
+   double ComputeNEffectiveEntries() const
+   {
+      if (fSumW2 == 0) {
+         return 0;
+      }
+      return fSumW * fSumW / fSumW2;
+   }
+
+   /// Compute the arithmetic mean of unbinned values.
+   ///
+   /// \f[
+   /// \mu = \frac{\sum w_i \cdot x_i}{\sum w_i}
+   /// \f]
+   ///
+   /// \param[in] dim the dimension index, starting at 0
+   /// \return the arithmetic mean of unbinned values
+   double ComputeMean(std::size_t dim = 0) const
+   {
```

Comment:
Is it planned to have Kahan sum correction, as with RDF.Mean?

Reply:
Currently not planned, unless requested for a clear use case.